DRILL-7972: ClassCastException when joining RDBMS and Elasticsearch tables

vvysotskyi opened the pull request:
# [DRILL-7972](https://issues.apache.org/jira/browse/DRILL-7972): ClassCastException when joining RDBMS and Elasticsearch tables

## Description
Updated `JdbcIntermediatePrelConverterRule` and `EnumerableIntermediatePrelConverterRule` to consider the convention of the input rel node when applying the rule, so now `EnumerableIntermediatePrelConverterRule` cannot be applied to `VertexDrel` rel node that has input with JDBC convention and vise versa.

Improved `FragmentContextImpl` to allow returning full root schema for both root and non-root fragments, since schema may be used when reading "enumerable" plugins.

## Documentation
NA

## Testing
Checked manually, no unit test added since the issue may be reproduced only with two different storage plugins.

paul-rogers reviewed (commented):
Seems OK, but I'm not very familiar with the Calcite part of Drill.

Would really like to see a unit test. There is no problem with setting up two plugins. Given the addition of traits, and `matches()`, would like to see tests that hit both the "matches" and "does not match" path.

In my limited experience, mental testing tends to not work well with Calcite: there are too many moving parts. Actual tests tend to turn up surprises.

vvysotskyi commented:
@paul-rogers, thanks for the review!
These plugins should be of different types to reproduce the issue, for example, JDBC and Elasticsearch, or JDBC and Cassandra, so it is a problem for adding end-to-end test since we don't have a module with several plugin dependencies.

Regarding updated rules, the main change is in its constructor, now it has the `RelOptRuleOperand` that controls conventions of the rel nodes, but it is slightly problematic to create all required instances for checking whether the rule would or wouldn't match the specific rel node.

paul-rogers commented:
@vvysotskyi, as it turns out, it is easy to create a "dummy" plugin for such cases. In fact, we have some. The "Mock" storage plugin (which is actually rather complex, because it also generates data). Another is this [super simple](https://github.com/apache/drill/blob/master/exec/java-exec/src/test/java/org/apache/drill/exec/store/BasePluginRegistryTest.java#L116) one.

What you may mean is that we don't have multiple plugins which we can use in an exec unit test, that need the rules which you are modifying. This does, indeed, make this kind of problem very hard to debug, and very hard to test. It also makes it very hard for anyone in the future to avoid breaking things.

Drill does not have "integration" tests, but it could: tests that are run after all plugins are created so the tests have visibility to the entire system. Presumably each plugin has some way to test itself (creates a test DB or whatever.) So, the integration test could use these features to set up two plugins. Very heavy-weight, and very hard to debug (still have to build everything.)

Point is: we do want to have tests so that people can confidently modify the code knowing that tests will catch any mistakes. How can we test fixes such as this?

paul-rogers commented:
Another thought. In years past, cases such as this were handled by the separate [Drill test framework](https://github.com/mapr/drill-test-framework) which was run before each release. The framework had visibility to all of Drill and a number of data sources as well. For this bug, it might be easier to add a test to that framework -- if we had a way to run the thing.

vvysotskyi commented:
@paul-rogers, I have added a mock plugin that can be used for checking this functionality and added unit tests that check both updated rules, please take a look.

paul-rogers reviewed (approved):
@vvysotskyi, thanks much for finding a way to test the change! This points the way for testing similar changes.

I am not a Calcite expert, so I did not review that aspect in detail. Otherwise, LGTM.
+1
